cpp/src/arrow/engine/key_map.h

```diff
@@ -0,0 +1,167 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <functional>
+
+#include "arrow/engine/util.h"
+#include "arrow/memory_pool.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+
+namespace arrow {
+namespace compute {
+
+//
+// 0 byte - 7 bucket | 1. byte - 6 bucket | ...
+// ---------------------------------------------------
+// |     Empty bit*   |    Empty bit       |
+// ---------------------------------------------------
+// |   7-bit hash    |    7-bit hash      |
+// ---------------------------------------------------
+// * Empty bucket has value 0x80. Non-empty bucket has highest bit set to 0.
+// ** The order of bytes is reversed - highest byte represents 0th bucket.
+// No other part of data structure uses this reversed order.
+//
+class SwissTable {
```

Comment:
Please add a higher level doc comment describing this class and it's utility. Specifically: describe how equality comparison and appending/storage of new entries are deferred to callbacks.

Separately, add a comment detailing its implementation and usage of stamps for vectorized probing.